### ISSUE TITLE
Word Embedding(흐름 + 실습) - [패캠]김기현의 딥러닝을 활용한 자연어처리

### ISSUE THREAD

**Issue by sallyy1** [sallyy1/NLU-NLP#5]
# word sense (단어의 의미)
- 유사성과 모호성
-언어의 성격: 효율성을 극대화하는 방향으로 진화=>생략, 시간에 따라 변화=>신조어 등
-단어의 "표제어(lemma)" 안의 여러 의미
1.동형어와 다의어 => "단어 중의성 해소(WSD, Word-sense disambiguation)" 필요 O
(ex. (traditional)주변 문맥을 통해 원래 단어의 의미를 파악하는 방법, (deeplearning)end-to-end 방법 및 RNN)
2.동의어 ("동의어 집합(Synset)")
3.상위어("hypernym")와 하위어("hyponym") -Hierarchy한 구조 => 어휘분류(taxonomy)
4.모호성 - 불연속적 심볼(겉) 속 내포된 의미를 찾기 위해 제거 필요

- Word의 성격: Discrete, not Continuous
-**discrete symbol** & **categorical value** 형태
-**불연속적 심볼** (이산 확률 변수)
-따라서, _연속적 값으로 표현_ 필요 (수치화, 벡터화, 워드임베딩 등)

## 1. 원핫인코딩
-단어를 컴퓨터가 인지할 수 있는 수치로 바꾸는 가장 간단한 벡터 표현 방법 (이산확률분포 => 멀티눌리 확률분포)
-원핫인코딩 벡터의 차원: Vocab Size
-(단점)0이 많은 sparse한 vector => 1)유사도 연산시 직교 많아 0결과값 많음 & 2)차원의 저주(: 같은 정보를 표현하는 데 필요한 공간 크기가 지수적으로 증가)

## 2. 시소러스를 활용한 단어 의미 파악 - (대표기법)Wordnet
-Thesaurus(어휘분류사전)
-Wordnet: NLTK 툴킷 임포트 or http://wordnetweb.princeton.edu/perl/webwn
-한국어 워드넷: KorLex, Korean WordNet(KWN)
-(장점) corpus가 없어도 & training set이 없어도 단어간 유사도 구할 수 있음
<img width="400" alt="스크린샷 2021-01-16 오후 6 25 09" src="https://user-images.githubusercontent.com/70500841/104808239-32412b80-5828-11eb-819d-52ad24687f50.png">
-(한계)
1.특정 Domain(또는 수집된 Data)에 특화된 수치를 계산하고 싶은 경우
2.신조어 or 사전에 등록되지 않은 단어
3.단어의 의미가 과거와 다르게 변화한 경우 등
사전에 대한 의존도가 높으므로, 활용도가 떨어질 수 있음 => **"Data-driven 방식의 필요성"**

---------------------------------------
* Data-driven 방식
: (장점)데이터가 충분하다면, **task**에 특화된 활용이 가능. => 특정 **Domain**/**Corpus**에 특화된 표현 가능.
(한계)1)여전히 sparse한 vector로 표현됨(PCA등 차원축소 가능) & 2)가정을 통해 설정한 hand craft한 feature

# (traditional) 딥러닝 이전의 단어 임베딩
- 자연어 처리에서 단어의 특징 벡터(feature vector) 구성을 위한 가정
1.의미가 비슷한 단어라면 쓰임새가 비슷할 것
2.쓰임새가 비슷하므로, 비슷한 문장 안에서 비슷한 역할로 사용될 것
3.따라서 함께 나타나는 단어들이 유사할 것

## 1. TF-IDF
- TF: Term Frequency, 단어의  문서 내에 출현한 횟수
X (일반적으로 많이 쓰이는 단어('the')에 패널티를 주는 효과)
- IDF : Inverse Document Frequency, 해당 단어가 출현한 문서 수의 역수
(한계)문서의 수가 너무 많다면 0이 많은 희소성 문제 O

## 2. Based on Context Window (Co-occurrence)
- 서로 유사한 단어들이 유사한 형태의 벡터를 갖도록 => windowing("컨텍스트 윈도우") -사용자가 값을 지정해주는 '하이퍼파라미터'
: 단어별로 윈도우 내에 속해 있는 이웃 단어들의 출현 빈도를 세어 행렬로 나타냄. (적절한 윈도우 크기 정하는 것이 중요)

--------------------------------------------------
* 벡터 유사도 구하기
1.L1 거리 (맨해튼 거리)
<img width="599" alt="스크린샷 2021-01-16 오후 6 37 52" src="https://user-images.githubusercontent.com/70500841/104808512-1e96c480-582a-11eb-9501-44d290928ce5.png">
2.L2거리 (유클리디안 거리)
<img width="618" alt="스크린샷 2021-01-16 오후 6 37 58" src="https://user-images.githubusercontent.com/70500841/104808507-163e8980-582a-11eb-858f-d7a4ff376646.png">
3.Infinity Norm
<img width="870" alt="스크린샷 2021-01-16 오후 6 38 08" src="https://user-images.githubusercontent.com/70500841/104808502-12126c00-582a-11eb-9669-87d7644b9def.png">
<img width="495" alt="스크린샷 2021-01-16 오후 6 41 17" src="https://user-images.githubusercontent.com/70500841/104808553-759c9980-582a-11eb-9803-c4576bcb3ae2.png">

4.코사인 유사도 : **방향**과 크기 모두 고려
<img width="460" alt="스크린샷 2021-01-16 오후 6 38 15" src="https://user-images.githubusercontent.com/70500841/104808498-0d4db800-582a-11eb-919d-e145cf3ab3f2.png">
5.자카드 유사도

-L1, L2 Norm과 Infinity Norm은 강조하고자 하는 것에 따라 사용
-Cosine Similarity는 벡터의 방향을 중요시 함. => Feature vector의 각 차원의 상대적인 크기가 중요할 때 사용.

--------------------------------------
# (딥러닝) 단어 임베딩

## 1. word2vec
- 1) CBOW : 주변에 나타나는 단어들(원핫인코딩 벡터) -> 해당 단어 1개 예측
- 2) skip-gram : 대상 단어(원핫인코딩 벡터) 1개 -> 주변에 나타나는 단어들 예측

- "Classification"을 통해 워드임베딩벡터(hidden vector)를 배울 수 있다. (Softmax)

## 2. GloVe
- 코퍼스에서 단어 X와 윈도우 내에 함께 출현한 (모든)단어들의 출현빈도를 맞추도록 훈련, "Regression" (MSE)
-(장점) 1) skip-gram보다 학습이 빠르다. (전체 corpus에 대해 각 단어별 co-occurrence를 구한 후, regression 수행) & 2) 출현빈도가 적은 단어들에 대해서는 loss의 기여도를 낮춤으로써, skip-gram의 단점 보완
-(한계) skip-gram과 큰 성능 차이 X

## 3. FastText
- Sum of subword embedding
- (장점) 빠르다. & 텍스트 분류를 훈련하는 기능도 O

--------------------------------------
- Word Embedding Vector(한계) 
-사전 훈련된 단어 임베딩 벡터는 해결하고자 하는 Task 문제에는 최적의 벡터 임베딩이 X. 목적함수가 다른 경우가 대부분.
-Word2vec의 목적: 주변단어를 잘 예측하도록 Feature를 구성하는 것
-Task 문제의 목적: 텍스트 분류, 언어모델, 번역 등..
-end-to-end 시스템을 지향하는 딥러닝에서 중간산출물이지, 최종산출물이 되는 Task 경우가 거의 없음

- Embedding Layer
-원핫벡터에서 해당 index row만 빠르게 가져오는 "Computer Efficiency" 역할. 원핫벡터를 Task에 맞는 dense한 표현으로 바꾸는 것 가능. 

--------------------------------------
# WordNet (Ch 04. Word Embedding - 04. 실습 WordNet을 활용한 단어 유사도 계산)

- wordnet.ipynb

# 딥러닝 이전의 단어 임베딩 (Ch 04. Word Embedding - 07. 실습 딥러닝 이전의 단어 임베딩 구현하기)

- word_similarity.ipynb

# FastText (Ch 04. Word Embedding - 13. 실습 Word Embedding)

- word_embedding_vector_exercise.ipynb
